Maintain an idSet property within QueryEmitter

distracteddev opened the pull request:
### Changes
- Update QueryEmitter to update an idSet property that maintains the same data as the `.ids` array, but in the form of a ES6 `Set()`. This was done as an optimization to avoid calling indexOf on potentially large arrays

### Notes
- This optimization is most effective in cases where queries return millions of documents
- I have confirmed that the current unit tests cover all branches of logic added/changed.

ericyhwang committed:
[Docs] Indicate create*Query options.results is a Doc[]

The `Connection`'s `createFetchQuery` and `createSubscribeQuery` take an optional existing result array as `options.results`.

It should be an array of Doc instances:
https://github.com/share/sharedb/blob/1cca122c63329665ebfdeceb1984921badb0cf4d/lib/client/connection.js#L563-L567

This PR updates the docs to indicate the specific type, instead of `Object[]`.

curran committed:
Fix typos in query docs

alecgibson committed:
📝 Remove `id` from `getOpsBulk()` documentation

distracteddev committed:
Maintain an idSet property within QueryEmitter as an optimization to avoid calling indexOf on potentially large arrays

This optimization is most effective on cases where queries can return millions of documents.

ericyhwang reviewed (changes requested):
The set needs to be re-created or otherwise updated here too:
https://github.com/lever/sharedb/blob/3d00dd73c57e1d096d5a7a8f4f806b23462106e5/lib/query-emitter.js#L202

distracteddev commented:
@ericyhwang argh. Great catch. Thank you!

distracteddev committed:
Recreate QueryEmitter.idSet when QueryEmitter.ids is redefined.

distracteddev commented:
@ericyhwang Updated

distracteddev commented:
Hrmm.. actually recreating the entire set is not exactly cheap if the `ids` is large enough.

Perhaps I should actually add the members of the `idsDiff` to the set iteratively.

distracteddev committed:
Add assertions to ensure QueryEmitter.ids matches QueryEmitter.idSet.

distracteddev committed:
Iteratively update QueryEmitter.set() instead of regenerating new Set within for every queryPoll.

distracteddev commented:
@ericyhwang I added some assertions that now correctly verify that the idSet matches the ids array in all modified branches of logic

With my most recent commit, I also update the queryPoll behavior to iteratively update the `idSet` instead of regenerating it entirely. Not sure its worth the extra code, would appreciate your perspective

distracteddev commented on lib/query-emitter.js:
```diff
@@ -311,16 +318,34 @@ QueryEmitter.prototype.onOp = function() {
+  return removedValues
+}
+
+function getDiffItemsByType(diff, diffType) {
```
tried to make this generic but then ended up needing specific logic for each diffType. Will likely refactor this so that `getRemoved()` and `getInserted()` don't share any code.

distracteddev committed:
Refactor getDiffItemsByType to be more effecient and avoid iterating over the same array twice, once for each diffType.

Also replaces an Array.concat with Array.push, since the latter is much more performant.

distracteddev committed:
Remove overly paranoid duck-type checks

distracteddev commented on lib/query-emitter.js:
```diff
@@ -310,17 +318,23 @@ QueryEmitter.prototype.onOp = function() {
         inserted.push(item.values[j]);
       }
+    } else if (item instanceof arraydiff.RemoveDiff) {
+      var values = currentValues.slice(item.index, item.index + item.howMany)
```
Should I validate that `item.index` and `item.howMany` are populated with the correct types?

ericyhwang replied:
If you mean, should we check `typeof foo === 'number'`, I don't think that's necessary, I don't recall needing to type-check arraydiff library results anywhere else.

curran committed:
Merge branch 'share:master' into patch-16

distracteddev commented:
Going to go ahead with this version and see how it performs in production

distracteddev committed:
Remove unused beforeEach in query-subscribe tests.

ericyhwang committed:
Merge pull request #555 from curran/patch-16

Fix typos in query docs

alecgibson committed:
Merge pull request #558 from share/get-ops-bulk-docs

📝 Remove `id` from `getOpsBulk()` documentation

ericyhwang committed:
More detail on create*Query prior results

Co-authored-by: Curran Kelleher <68416+curran@users.noreply.github.com>

alecgibson committed:
Merge pull request #546 from share/docs-createQuery-options-types

[Docs] Indicate create*Query options.results is a Doc[]

alecgibson committed:
💥 Update Node.js test matrix

Node.js 18 [has been released][1], and will be the next LTS version, so
we add this to our test matrix.

Node.js 12 has also been [end-of-lifed][2], and is dropped from the test
matrix.

[1]: https://nodejs.org/en/blog/announcements/v18-release-announce/
[2]: https://nodejs.org/en/about/releases/

dependabot[bot] committed:
Bump nokogiri from 1.13.3 to 1.13.4 in /docs

Bumps [nokogiri](https://github.com/sparklemotion/nokogiri) from 1.13.3 to 1.13.4.
- [Release notes](https://github.com/sparklemotion/nokogiri/releases)
- [Changelog](https://github.com/sparklemotion/nokogiri/blob/main/CHANGELOG.md)
- [Commits](https://github.com/sparklemotion/nokogiri/compare/v1.13.3...v1.13.4)

---
updated-dependencies:
- dependency-name: nokogiri
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>

alecgibson committed:
Merge pull request #563 from share/dependabot/bundler/docs/nokogiri-1.13.4

Bump nokogiri from 1.13.3 to 1.13.4 in /docs

alecgibson committed:
Merge pull request #562 from share/node-18

💥 Update Node.js test matrix

alecgibson committed:
3.0.0

alecgibson committed:
📝 Update presence documentation

This change updates our presence documentation to highlight that
disconnections can be seen with a `null` value.

alecgibson committed:
👷 Bump GH actions

alecgibson committed:
👷 Verbose GH Actions installation

alecgibson committed:
👷 Increase GH Actions timeout

The Node.js v14 `npm install` script keeps stalling. This change
increases our timeout to try and let it pass.

alecgibson committed:
👷 Use HTTPS to fetch `json0`

The older version of `npm` can have issues installing packages directly
from Github over SSH. This change updates our `package.json` to
explicitly fetch over HTTPS.

alecgibson committed:
👷 Fix `json0` URL

alecgibson committed:
Revert "👷 Increase GH Actions timeout"

This reverts commit e71c91b88f69a2f49994141a9688a662d41e6612.

alecgibson committed:
Revert "👷 Verbose GH Actions installation"

This reverts commit 16d7431c6d5314da3e4e5215721379618556b086.

alecgibson committed:
📌 Pin `json0` dev dependency

It's best practice to specify a version when depending directly on a
repo.

dependabot[bot] committed:
Bump nokogiri from 1.13.4 to 1.13.7 in /docs

Bumps [nokogiri](https://github.com/sparklemotion/nokogiri) from 1.13.4 to 1.13.7.
- [Release notes](https://github.com/sparklemotion/nokogiri/releases)
- [Changelog](https://github.com/sparklemotion/nokogiri/blob/main/CHANGELOG.md)
- [Commits](https://github.com/sparklemotion/nokogiri/compare/v1.13.4...v1.13.7)

---
updated-dependencies:
- dependency-name: nokogiri
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>

alecgibson committed:
Merge pull request #567 from share/pin-json0

📌 Pin `json0` dev dependency

alecgibson committed:
Merge pull request #568 from share/dependabot/bundler/docs/nokogiri-1.13.7

Bump nokogiri from 1.13.4 to 1.13.7 in /docs

alecgibson committed:
Merge pull request #566 from share/docs-null-presence

📝 Update presence documentation

dependabot[bot] committed:
Bump tzinfo from 1.2.9 to 1.2.10 in /docs

Bumps [tzinfo](https://github.com/tzinfo/tzinfo) from 1.2.9 to 1.2.10.
- [Release notes](https://github.com/tzinfo/tzinfo/releases)
- [Changelog](https://github.com/tzinfo/tzinfo/blob/master/CHANGES.md)
- [Commits](https://github.com/tzinfo/tzinfo/compare/v1.2.9...v1.2.10)

---
updated-dependencies:
- dependency-name: tzinfo
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>

alecgibson committed:
Merge pull request #569 from share/dependabot/bundler/docs/tzinfo-1.2.10

Bump tzinfo from 1.2.9 to 1.2.10 in /docs

ericyhwang committed:
Refactor action code strings into an enum-like object

ericyhwang committed:
Updates to message-actios from code review feedback

ericyhwang committed:
Update action refactoring missed in automated renames

ericyhwang committed:
Rename queryReply to queryUpdate, use shared action object from query.js

alecgibson committed:
Catch missed actions

ericyhwang committed:
Add top-level export for MESSAGE_ACTIONS

ericyhwang committed:
Merge pull request #570 from share/refactor-action-codes

Refactor action code strings into an enum-like object

ericyhwang committed:
Refactor message action codes into a shared file

alecgibson committed:
✨ Add ping/pong

This change adds the ability for a client to send a "ping" to the server
and receive a "pong" response.

The motivation for this is that sometimes the socket underlying the
`Connection` may not reliably report its connection state (for example
if using a wrapper around a "vanilla" websocket that handles
reconnection).

The most bullet-proof way for a client to determine its connection state
is to actually make a request to the server and assert that it receives
a timely response.

The implementation of ping/pong is arguably a websocket concern,
especially since it's already [defined by the spec][1]. However, the
browser JavaScript [`WebSocket`][2] does not expose this functionality,
so we have to add our own ping/pong layer on top anyway.

It's also worth noting that consumers of this library can't easily
send their own ping messages down the socket, since ShareDB will
[error][3] for unknown messages.

Note that this change only adds the ability for a client to ping the
server, and not the other way around. This is because:

 - the `Agent` is not an `Emitter`, and emitting a `pong` on the
   `Backend` is pretty meaningless
 - server-side implementations of WebSockets, such as `ws`, _do_
   [expose a `ping` API][4]

[1]: https://www.rfc-editor.org/rfc/rfc6455#section-5.5.2
[2]: https://developer.mozilla.org/en-US/docs/Web/API/WebSocket
[3]: https://github.com/share/sharedb/blob/8b531bedf19860ffe0b37a3e1c8da7a32b5005bd/lib/agent.js#L451
[4]: https://github.com/websockets/ws/blob/966f9d47cd0ff5aa9db0b2aa262f9819d3f4d414/lib/websocket.js#L351

alecgibson committed:
Merge pull request #576 from share/ping-pong

✨ Add ping/pong

alecgibson committed:
3.2.0

dependabot[bot] committed:
Bump commonmarker from 0.23.4 to 0.23.6 in /docs

Bumps [commonmarker](https://github.com/gjtorikian/commonmarker) from 0.23.4 to 0.23.6.
- [Release notes](https://github.com/gjtorikian/commonmarker/releases)
- [Changelog](https://github.com/gjtorikian/commonmarker/blob/main/CHANGELOG.md)
- [Commits](https://github.com/gjtorikian/commonmarker/compare/v0.23.4...v0.23.6)

---
updated-dependencies:
- dependency-name: commonmarker
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>

alecgibson committed:
📝 Add `connection.ping()` to the documentation

alecgibson committed:
Merge pull request #577 from share/dependabot/bundler/docs/commonmarker-0.23.6

Bump commonmarker from 0.23.4 to 0.23.6 in /docs

alecgibson committed:
Merge pull request #578 from share/ping-pong-docs

📝 Add `connection.ping()` to the documentation

alecgibson committed:
🐛 Allow getting `Presence` while it's being destroyed

There's currently an edge case involving trying to get presence in the
middle of being destroyed. For example, consider these synchronous
calls:

```js
var presence = connection.getPresence('channel')
presence.destroy()
var newPresence = connection.getPresence('channel')
```

Since `Presence` objects are per-channel singletons, the above code is
effectively the same as:

```js
var presence = connection.getPresence('channel')
presence.destroy()
var newPresence = presence
```

In other words, our "new" presence is actually already being destroyed.
What's worse, is that if we start trying to create local presence before
the `destroy()` callback is triggered, that gets "cleaned up" by the
initial call to `destroy()`:

```js
var presence = connection.getPresence('channel')
presence.destroy(() => {
  // by the time we reach here, my newly-created localPresence
  // will have been destroyed
})
var newPresence = connection.getPresence('channel')
var localPresence = newPresence.create('me')
```

This change attempts to address this edge case by adding a private
`_wantsDestroy` to our `Presence` instances, which is:

  - `false` on construction
  - set to `true` when calling `destroy()`
  - reset to `false` if calling `connection.getPresence()`

We then check this flag when performing the `destroy()` tidy-up, and
avoid tidying up if we have since reset the flag.

Note that there's a corner case within this edge case, surrounding
`LocalPresence` instances created while a `destroy()` is in progress.
Since we might create `LocalPresence` instances at any point during the
asynchronous `destroy()`, we define the following behaviour:

  - any `LocalPresence` instances that exist at the time of invoking
    `destroy()` will be destroyed, **even if** we subsequently cancel
    the destroy (by invoking `connection.getPresence()`)
  - any `LocalPresence` instances that are created *after* invoking
    `destroy()` will be kept if the destroy is cancelled
  - creating a new `LocalPresence` while `_wantDestroy === true` will
    throw, because clients shouldn't put themselves in this situation
    and the behaviour when tidying this up is undefined

alecgibson committed:
🐛 Allow getting a `Doc` while it's being destroyed

There's currently an edge case involving trying to get a doc in the
middle of being destroyed. For example, consider these synchronous
calls:

```js
var doc = connection.get('collection', 'id')
doc.destroy()
var newDoc = connection.get('collection', 'id')
```

Since `Doc` objects are singletons, the above code is effectively the
same as:

```js
var doc = connection.get('collection', 'id')
doc.destroy()
var newDoc = doc
```

In other words, our "new" doc is actually already being destroyed, and
will soon be removed from the `Connection` cache.

One of the ways this could cause a bug is that the `doc` cannot be found
in order to trigger callbacks when receiving responses from the server.
`Connection` already tries to handle this case with [`_addDoc()`][1],
but this could be problematic for a couple of reasons:

 1. If we add new functionality to `Connection` we may forget to add a
    call to `_addDoc()`
 2. This is susceptible to having multiple different instances of a
    `Doc` continuously trying to replace one another on the `Connection`
    cache

We can reach multiple `Doc`s trying to replace one another with the
following edge case:

 1. Get a `Doc`
 2. Destroy it
 3. Synchronously get the same `Doc` (returning the instance in the
    process of being destroyed)
 4. Wait for the `destroy()` to finish (ie the `Doc` has been removed
    from the `Connection` cache)
 5. Get the `Doc` again - we now have a fresh, different instance
 6. Try to submit ops on both docs at the same time
 7. Depending on which one goes last, that one will be the `Doc` stored
    on the `Connection`, and will incorrectly receive both server
    responses

This change attempts to address this edge case by adding a private
`_wantsDestroy` to our `Doc` instances, which is:

  - `false` on construction
  - set to `true` when calling `destroy()`
  - reset to `false` if calling `connection.get()`

We then check this flag when performing the `destroy()` tidy-up, and
avoid tidying up if we have since reset the flag.

The end result of this is that any calls to `connection.get()` before
`destroy()` has resolved will return the same singleton `Doc` instance,
and prevent tidy-up.

[1]: https://github.com/share/sharedb/blob/5176283e2d1c9dc263f1fa3c2bc6a93c1afe2ad3/lib/client/connection.js#L452-L453

alecgibson committed:
Merge pull request #580 from share/destroying-docs

🐛 Allow getting a `Doc` while it's being destroyed

alecgibson committed:
Merge pull request #579 from share/destroying-presence

🐛 Allow getting `Presence` while it's being destroyed

alecgibson committed:
3.2.1

alecgibson committed:
✨ Allow middleware to mutate submitted ops

This change adds a new `request.$fixup()` method to the `SubmitRequest`
class, which allows middleware to mutate ops in the `apply` middleware.

Motivation
----------
There are times when we want the server to make changes to ops submitted
by a client. There may be many reasons for wanting to do this, including
distrusting a client (eg setting timestamp, or auth information);
keeping sensitive business logic out of client-side code; etc.

The only ways of doing this currently are:

 1. Committing the client's op, and then submitting an amendment op from
    the server; or
 2. Rejecting the client's op, and submitting the full amended op from
    the server

Option 1 works, but possibly commits badly-formed, or untrustworthy data
to the database, and broadcasts it to remote clients. It also makes it
possible to fetch historic snapshots of this untrustworthy in-between
state, which undermines the trustworthiness of our data.

Option 2 also works, and might work with invertible types through a
"soft" rollback, but can lose pending ops if we're not careful, or if
the type isn't invertible.

Approach
--------
This change adds a `$fixup()` helper function to the `SubmitRequest`
object, named after the `git` action for performing similar actions.

The idea of the `$fixup()` function is to act as a middleware-friendly
`submitRequest()`.

Any ops submitted through `$fixup()` will be:

 - composed with the `request.op`, so downstream middleware can work on
   the fixed-up op
 - sent back to the original client, and applied
 - help to avoid infinite middleware loops, since we don't trigger any
   nested submission calls

Note that in order to handle the case of a client disconnecting before
getting their fixup, we now also store the fixup ops on the op `m`
metadata field, and fetch this when getting ops to the current snapshot.

Limitations
-----------
This feature will only work:

 - on types which implement the optional `.compose()` function
 - in the `apply` middleware, where the snapshot's `type` is available,
   but before the op has actually been applied to the snapshot
 - on creation or edit ops (not deletion)

alecgibson committed:
👷‍♀️ Move to `ruby/setup-ruby` action

The [`actions/setup-ruby`][1] action has been deprecated and advises
moving to [`ruby/setup-ruby`][2]

[1]: https://github.com/actions/setup-ruby
[2]: https://github.com/ruby/setup-ruby

alecgibson committed:
Merge pull request #582 from share/bump-actions-versions

👷‍♀️ Move to `ruby/setup-ruby` action

dependabot[bot] committed:
Bump nokogiri from 1.13.7 to 1.13.9 in /docs

Bumps [nokogiri](https://github.com/sparklemotion/nokogiri) from 1.13.7 to 1.13.9.
- [Release notes](https://github.com/sparklemotion/nokogiri/releases)
- [Changelog](https://github.com/sparklemotion/nokogiri/blob/main/CHANGELOG.md)
- [Commits](https://github.com/sparklemotion/nokogiri/compare/v1.13.7...v1.13.9)

---
updated-dependencies:
- dependency-name: nokogiri
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>

alecgibson committed:
Merge pull request #583 from share/dependabot/bundler/docs/nokogiri-1.13.9

Bump nokogiri from 1.13.7 to 1.13.9 in /docs

Dawidpol committed:
✨ Add multi channel query subscription.

At the moment we can only specify one channel the specific query, would be able to listen to, as per docs:
```javascript
backend.use('query', (context, next) => {
  // Set our query to only listen for changes on our user-specific channel
  context.channel = userChannel(context)
  next()
})
```

However let's imagine the situation where the user wants to query all
the posts, the where posted by him and all his friends. Now we would need
new query every for friend separately to make sure the proper scalability is
preserved and we do not receive all the changes to posts collection. This change allows to listen for multiple channels, so if we
want to query all user friends posts. We can do it by this:
```javascript
backend.use('query', (context, next) => {
  // Set our query to only listen for changes on our user-specific channel
  context.channels = [userChannel(context), friendChannel(context))]
  next()
})
```

Now this query would only listen to all the changes that were made to the user
posts and his friends.

alecgibson committed:
🔇 Fix sensitive memory leak warnings when using presence

Fixes https://github.com/share/sharedb/issues/584

At the moment, we [attach `Doc` listeners][1] for every remote client in
a document (as well as every local user).

This means `Doc`s can quickly hit the [default of 10][2] maximum event
listeners.

This change adds a `DocPresenceEmitter` class, which only registers any
given `Doc` listener once, regardless of how many presence instances
there are. It then forwards events on through its own emitters, which
have a much higher value of 1000 max listeners set, in order to avoid
too much alert spam, but also keep *some* checking for memory leaks.

[1]: https://github.com/share/sharedb/blob/cc0e3382bf3df4c38c7ccc6bc52da71aff8c9f74/lib/client/presence/remote-doc-presence.js#L42-L47
[2]: https://nodejs.org/api/events.html#eventsdefaultmaxlisteners

alecgibson committed:
Merge pull request #588 from share/presence-events

🔇 Fix sensitive memory leak warnings when using presence

alecgibson committed:
3.2.2

alecgibson committed:
✅ Fix flaky tests

Fixes https://github.com/share/sharedb-mongo/issues/131

This change attempts to fix some tests that are flaky in `sharedb-mongo`.

The flakiness can be reproduced locally by wrapping the
`Agent._querySubscribe()` [call to `_fetchBulkOps()`][1] in a long
`setTimeout()`:

```js
  if (options.fetchOps) {
    wait++;
    setTimeout(function() {
      console.log('fetch bulk ops');
      agent._fetchBulkOps(collection, options.fetchOps, finish);
    }, 1000);
  }
```

This forces us into an edge case where the subscribe query triggers and
returns the diff from a [`queryPoll()`][2], which triggers the tests'
`'insert'` handlers, finishes the test, and closes the backend, all
before the `_fetchBulkOps()` call is issued (which subsequently fails
because the database has been closed).

Handling this query subscribe actually triggers a variety of responses
to be sent to the client at different times:

 1. The actual `_querySubscribe()` [callback][3] (which ultimately
    triggers [`agent._reply()`][4] in response to the original request)
 2. Ops sent [independently][5] by [`_fetchBulkOps()`][6]
 3. The diff resulting from [`queryPoll()`][2]

In order to reduce flakiness, this change adds checks that the query's
[`'ready'` event][7] has been called, which will happen once the
resubscribe request has been replied to by the `Agent`.

This ensures we've waited for events 1 & 3 of the above list, although
we notably aren't waiting for event 2 (which is where the error is
actually coming from).

Since no ops will actually be sent to the client, I'm not sure how best
to wait for this. Hopefully waiting for the subscribe ack should be
sufficient.

[1]: https://github.com/share/sharedb/blob/b84290f8a66ee42b3d2c75d6de780e125b99c514/lib/agent.js#L521-L524
[2]: https://github.com/share/sharedb/blob/b84290f8a66ee42b3d2c75d6de780e125b99c514/lib/agent.js#L534
[3]: https://github.com/share/sharedb/blob/b84290f8a66ee42b3d2c75d6de780e125b99c514/lib/agent.js#L511
[4]: https://github.com/share/sharedb/blob/b84290f8a66ee42b3d2c75d6de780e125b99c514/lib/agent.js#L344
[5]: https://github.com/share/sharedb/blob/b84290f8a66ee42b3d2c75d6de780e125b99c514/lib/agent.js#L265
[6]: https://github.com/share/sharedb/blob/b84290f8a66ee42b3d2c75d6de780e125b99c514/lib/agent.js#L523
[7]: https://github.com/share/sharedb/blob/b84290f8a66ee42b3d2c75d6de780e125b99c514/lib/client/query.js#L148

alecgibson committed:
Merge pull request #590 from share/flaky-query-test-fix

✅ Fix flaky tests

alecgibson committed:
3.2.3

jameshfisher committed:
Add missing import in hello world example

ericyhwang committed:
Merge pull request #591 from jameshfisher/patch-1

Add missing import in hello world example

alecgibson committed:
✅ Remove order dependence from test

alecgibson committed:
Merge pull request #592 from share/fix-test

✅ Remove order dependence from test

alecgibson committed:
3.2.4

curran committed:
Upgrade deps in examples. Closes #594

curran committed:
Fix leaderboard example

curran committed:
Fix rich-text-presence example

curran committed:
Start on Vite example

curran committed:
Fix CI

dawidreedsy committed:
Merge pull request #587 from share/feature/multi-channel-subscription

✨ Add multi channel query subscription.

alecgibson committed:
Merge pull request #581 from share/server-fixup

✨ Allow middleware to mutate submitted ops

alecgibson committed:
3.3.0

curran committed:
Got Vite example to work

curran committed:
Upgrade ShareDB in examples. Fix presence API warning.

curran committed:
Fix React warnings

alecgibson committed:
Merge pull request #595 from curran/upgrade-deps-in-examples

Upgrade dependencies in examples

alecgibson committed:
✅ Try to fix flaky Milestone DB tests

At the moment we sometimes get some flaky Milestone Mongo test
[failures][1].

This potentially happens because milestone snapshots are saved as a
[fire-and-forget request][2], which means that we technically make no
guarantee about the order that snapshots are written to the database.

Since this is the case, the affected tests may write the v4 snapshot
before the others, and then fail assertions about earlier snapshots.

This change tweaks these tests to count the snapshots instead, which
should be more robust to this race condition.

[1]: https://github.com/share/sharedb-milestone-mongo/actions/runs/4545270938/jobs/8012439274
[2]: https://github.com/share/sharedb/blob/404cde5568bf88d8de12186a90ba6bc168920e23/lib/submit-request.js#L223

alecgibson committed:
Merge pull request #602 from share/flaky-milestone-test-fix

✅ Try to fix flaky Milestone DB tests

alecgibson committed:
3.3.1

curran committed:
Use glob override for .eslintrc

curran committed:
Make it work

curran committed:
Lint

ericyhwang committed:
Merge pull request #598 from curran/counter-json1-vite

Vite example

alecgibson committed:
💥 Drop Node.js 14 and add Node.js 20

According to the Node.js [release schedule][1]:

 - v14 will be end-of-lifed on 30 April
 - v20 has been released

This change drops v14 from our test matrix, and adds v20.

[1]: https://nodejs.dev/en/about/releases/

alecgibson committed:
Merge pull request #606 from share/bump-node

💥 Drop Node.js 14 and add Node.js 20

alecgibson committed:
🔊 Warn when messages are sent before the v1.1 handshake

At the moment, it's possible for messages to be sent before the client-
server handshake.

Sending messages before the handshake has happened has undefined
behaviour, and can result in errors such as in:
https://github.com/share/sharedb/issues/605

We can't just ignore these messages, because old clients might
potentially be on v1.0 of the client-server protocol, in which the
server informs the client when it's ready, but not the other way around,
so it's impossible to know when a client should be considered "ready",
and its messages acceptable.

Instead, we add a warning for clients on v1.1 who have sent a message
before their handshake. In order to aid with debugging, we keep track of
the first message received, and log it when the handshake is received
(which means that v1.0 clients will never get such a warning).

alecgibson committed:
📝 Update docs around `reconnecting-websocket`

In our docs, we recommend using [`reconnecting-websocket`][1] to handle
websocket reconnection.

However, [by default][2] that library will [buffer][3] messages when its
underlying socket is closed, which can lead to [undefined behaviour][4]
when ShareDB reconnects, since it works under the assumption that all
messages sent as the socket is closing have been lost (eg pushing
inflight ops back onto the pending queue, etc.).

This change updates our documentation and our examples to set
`{maxEnqueuedMessages: 0}`, which disables this buffering, and should
help to avoid ShareDB reaching undefined states when reconnecting using
this library.

[1]: https://www.npmjs.com/package/reconnecting-websocket
[2]: https://github.com/pladaria/reconnecting-websocket/blob/05a2f7cb0e31f15dff5ff35ad53d07b1bec5e197/reconnecting-websocket.ts#L46
[3]: https://github.com/pladaria/reconnecting-websocket/blob/05a2f7cb0e31f15dff5ff35ad53d07b1bec5e197/reconnecting-websocket.ts#L260
[4]: https://github.com/share/sharedb/issues/605

alecgibson committed:
Merge pull request #607 from share/pre-handshake-warning

🔊 Warn when messages are sent before the v1.1 handshake

alecgibson committed:
Merge pull request #608 from share/reconnecting-socket-docs

📝 Update docs around `reconnecting-websocket`

alecgibson committed:
⚡️ Deduplicate `fetch()` network requests

At the moment, if a client calls `doc.fetch()` with a fetch request
already in progress, the client will send a second request, which
probably isn't needed (or wanted).

This change prevents duplicate requests from being sent when there's
already an inflight fetch request, and adds an internal `force` option,
which can be used to force a duplicate request to be sent. This is
currently used internally for `_hardRollback()`, which forcibly unsets
the `doc.version`, and therefore requires a separate `fetch()` call,
since fetch requests are [handled differently][1] depending on the
presence of `version`.

[1]: https://github.com/share/sharedb/blob/21750e320f6865ee14620d6bf9610cb7ebc5f18e/lib/agent.js#L607

alecgibson committed:
Merge pull request #610 from share/merge-inflight-fetch

⚡️ Deduplicate `fetch()` network requests

alecgibson committed:
4.0.0

curran committed:
Upgrade dependencies

curran committed:
Revert eslint

curran committed:
Upgrade eslint

curran committed:
Got lint to work

alecgibson committed:
Merge pull request #618 from curran/upgrade-deps-aug-2023

Upgrade dependencies except ESLint

curran committed:
Merge branch 'master' into upgrade-eslint

curran committed:
Restore deps ordering

alecgibson committed:
➖ Remove `lolex`

`sinon` has ships with its own fake timers, so let's just use those.

alecgibson committed:
Merge pull request #619 from curran/upgrade-eslint

Upgrade eslint

alecgibson committed:
Merge pull request #620 from share/remove-lolex

➖ Remove `lolex`

alecgibson committed:
🔥 Remove flaky test

This test is failing pretty consistently in [`sharedb-mongo`][1].

This test has notably had problems [in the past][2], so let's just
delete it.

[1]: https://github.com/share/sharedb-mongo/actions/runs/6013380406/job/16310733641?pr=151
[2]: https://github.com/share/sharedb/pull/324

alecgibson committed:
Merge pull request #621 from share/flaky-test-debug

🔥 Remove flaky test

alecgibson committed:
4.0.1

ericyhwang committed:
Remove extra = in sharedb-legacy devDependency

Fixes "Invalid URL" error in latest Node 18.17.1 with NPM 6

ericyhwang committed:
Merge pull request #623 from share/fix-sharedb-legacy-spec

Remove extra = in sharedb-legacy devDependency

ericyhwang committed:
4.0.2

ericyhwang committed:
Merge branch 'master' into email-agent-fork

dwinrick-lever commented:
@distracteddev @ericyhwang  should we consider merging this so everyone benefits? I think it's been battle tested for over a year with email-agent. The additional memory usage should be very tiny due to string interning.

Dawidpol committed:
♻  Invert if to reduce nesting

dawidreedsy committed:
Merge pull request #625 from share/refactor-code-invert-if

♻  Invert if to reduce nesting

alecgibson committed:
📈 Add `presence.broadcast` `timing` event

This change adds a `presence.broadcast` `timing` event, which can be
used to monitor Presence broadcast performance.

alecgibson committed:
⚡️ Avoid repeat subscription requests

At the moment, if a `Presence` instance is already subscribed, calling
`presence.subscribe()` will send another subscription request, which
can cause unnecessary server load, including:

 - [creating a new Pub/Sub subscription][1], and
 - [re-requesting all other presences][2]

This change updates our presence subscription logic to early return if
our desired subscription state already matches both `wantSubscribe`
**and** `subscribed`. We also reset `subscribed` to `false` when our
connection is disconnected to correctly update our state.

[1]: https://github.com/share/sharedb/blob/62e4ec5d46c0dcb097931e9dae60240aaba6b2b3/lib/agent.js#L811
[2]: https://github.com/share/sharedb/blob/62e4ec5d46c0dcb097931e9dae60240aaba6b2b3/lib/agent.js#L820

alecgibson committed:
⚡️ Fix Presence `Stream` leak

At the moment, when subscribing a `Presence` multiple times, multiple
PubSub `Stream`s are created per-subscription.

However, unsubscribing will only destroy the [last created][1],
resulting in streams that are never destroyed.

This change updates our subscription logic to check if we already have
an existing stream. If we already have a stream, then we just
re-request presence from other clients (which is needed by our
[`DocPresence` logic][2]), and return.

Note that we also tweak our unsubscribe logic to eagerly remove the
stream after calling destroy (so that subsequent subscribes don't try to
reuse a stream that is currently being destroyed).

We also remove a check against `seq`, which is currently not covered by
tests (and actually never calls the `callback`), so is clearly broken
and unneeded anyway.

[1]: https://github.com/share/sharedb/blob/62e4ec5d46c0dcb097931e9dae60240aaba6b2b3/lib/agent.js#L829-L830
[2]: https://github.com/share/sharedb/blob/62e4ec5d46c0dcb097931e9dae60240aaba6b2b3/lib/client/presence/remote-doc-presence.js#L108-L110

alecgibson committed:
Merge pull request #627 from share/presence-timing

📈 Add `presence.broadcast` `timing` event

alecgibson committed:
Merge pull request #628 from share/presence-subscribe-once

⚡️ Avoid repeat subscription requests

dawidreedsy committed:
♻  Throw error for pending ops on hard rollback (#626)

Our current flow for handling the rollback swallows some errors and discards pending ops.

1. Whenever we do hard rollback and fetch for hard rollback fails we should throw the error so that user is informed that something really wrong happen.
2. Whenever we preform the hard rollback and we have some pending ops and the `ERR_OP_PENDING_OP_SUBMIT_REJECTED` error is thrown we should soft reject, as this error is meant for the user to reject some of the ops from submition to DB, however we should throw for all the pending ops s the pending ops would be otherwise just discarded. This way we at least allow user to handle the error
3. For the soft rollback we just resolve for this error and allow pending ops to be submitted to the server.

alecgibson committed:
Merge pull request #629 from share/subscription-stream-leak

⚡️ Fix Presence `Stream` leak

alecgibson committed:
🏷️ Mark `presence.subscribe()` options as internal

Add a prefix to the `force` option to mark it as internal, so we can
remove it later if needed.

alecgibson committed:
Merge pull request #630 from share/internalise-force-subscribe

🏷️ Mark `presence.subscribe()` options as internal

alecgibson committed:
4.1.0

alecgibson committed:
♻️ Remove `presence.subscribe()` `force` option

We previously only needed this to re-request presence as part of the
`DocPresence` catchup flow.

This change adds an internal method to actively request remote presence
for a given channel, and uses that instead of abusing `subscribe()`.
This lets us remove the `options` object from `subscribe()`.

alecgibson committed:
Merge pull request #631 from share/request-presence

♻️ Remove `presence.subscribe()` `force` option

alecgibson committed:
4.1.1

alecgibson committed:
🐛 Prevent caching presence broadcasts rejected by middleware

The `Agent` stores the most recent presence broadcast sent by a client.
However, it eagerly stores this before running the `receivePresence`
middleware, which may actually decide that the presence broadcast is
invalid (eg malformed, unauthorized, etc.).

This means that the latest cached presence state may not be "legal".

This change moves the caching inside the `trigger()` call, so we only
store presence values that have "passed" the middleware.

ericyhwang committed:
Fix and guard against prototype pollution issues

ericyhwang committed:
Merge pull request #635 from share/fix-proto-clobber

Fix and guard against prototype pollution issues

alecgibson committed:
Merge pull request #634 from share/receive-presence-middleware

🐛 Prevent caching presence broadcasts rejected by middleware

alecgibson committed:
4.1.2

dependabot[bot] committed:
Bump commonmarker from 0.23.6 to 0.23.10 in /docs

Bumps [commonmarker](https://github.com/gjtorikian/commonmarker) from 0.23.6 to 0.23.10.
- [Release notes](https://github.com/gjtorikian/commonmarker/releases)
- [Changelog](https://github.com/gjtorikian/commonmarker/blob/v0.23.10/CHANGELOG.md)
- [Commits](https://github.com/gjtorikian/commonmarker/compare/v0.23.6...v0.23.10)

---
updated-dependencies:
- dependency-name: commonmarker
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
Bump nokogiri from 1.13.9 to 1.15.5 in /docs

Bumps [nokogiri](https://github.com/sparklemotion/nokogiri) from 1.13.9 to 1.15.5.
- [Release notes](https://github.com/sparklemotion/nokogiri/releases)
- [Changelog](https://github.com/sparklemotion/nokogiri/blob/main/CHANGELOG.md)
- [Commits](https://github.com/sparklemotion/nokogiri/compare/v1.13.9...v1.15.5)

---
updated-dependencies:
- dependency-name: nokogiri
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>

alecgibson committed:
Merge pull request #637 from share/dependabot/bundler/docs/nokogiri-1.15.5

Bump nokogiri from 1.13.9 to 1.15.5 in /docs

alecgibson committed:
Merge pull request #636 from share/dependabot/bundler/docs/commonmarker-0.23.10

Bump commonmarker from 0.23.6 to 0.23.10 in /docs

alecgibson committed:
⬆️ Bump GitHub pages gem

alecgibson committed:
Merge pull request #638 from share/bump-gh-pages-gem

⬆️ Bump GitHub pages gem

alecgibson committed:
👷‍♀️ Specify `bundler` version

The docs build currently fails with:

```
Run cd docs && gem install bundler && bundle install
ERROR:  Error installing bundler:
	The last version of bundler (>= 0) to support your Ruby & RubyGems was 2.4.22. Try installing it with `gem install bundler -v 2.4.22`
```

This change follows the advice of the error message, and specifies the
version of bundler to install, since we [can't bump Ruby][1].

[1]: https://pages.github.com/versions/

alecgibson committed:
Merge pull request #641 from share/pin-bundler

👷‍♀️ Specify `bundler` version

distracteddev committed:
Merge branch 'master' of github.com:lever/sharedb into email-agent-fork
